h/services/search_index/service.py

```diff
@@ -1,50 +1,88 @@
+from h import storage
 from h.events import AnnotationTransformEvent
 from h.presenters import AnnotationSearchIndexPresenter
 
 
 class SearchIndexService:
-    def __init__(self, es_client, request):
-        self.es_client = es_client
+    """Service for manipulating the search index."""
+
+    # The DB setting that stores whether a full re-index is taking place
+    REINDEX_SETTING_KEY = "reindex.new_index"
+
+    def __init__(self, request, es_client, session, settings):
+        """
+        Create an instance of the service.
+
+        :param request: Pyramid request object
+        :param es_client: Elasticsearch client
+        :param session: DB session
+        :param settings: Instance of settings (or other object with `get()`)
+        """
         self.request = request
+        self.es_client = es_client
+        self.session = session
+        self.settings = settings
+
+    def add_annotation_by_id(self, annotation_id):
+        """
+        Add an annotation into the search index by id.
 
-    def add_annotation(self, annotation, target_index=None):
+        A new annotation document will be created in the search index or,
+        if the index already contains an annotation document with the same Id
+        as the given annotation then it will be updated.
+
+        If no annotation is found, nothing happens.
+
+        :param annotation_id: Id of the annotation to add.
         """
-        Index an annotation into the search index.
+        annotation = storage.fetch_annotation(self.session, annotation_id)
+        if not annotation:
+            return
+
+        self.add_annotation(annotation)
+
+        if annotation.is_reply:
+            self.add_annotation_by_id(annotation.thread_root_id)
```

Comment:
This is a subtle change of behavior. Previously this would have been `delay`'ed out to another instance of the celery task. Now it's done synchronously here.